### DISCUSSION TITLE
Keyball44　PC起動時の動作・LED輝度について

### DISCUSSION THREAD

**Discussion by heppokoX** [Yowkees/keyball#194] [Q&A]
お世話になります。
先日Keyball44を購入させていただき、ひとまず使用できるようにはなったのですが
一部疑問点がありましたので投稿させていただきます。

①PC起動時に動作しない
Windows11環境で使用しておりますが、Keyball44を接続した状態でPCの電源を入れると
両側のOLEDにKeyballロゴが表示されキーボードもトラックボールも動作しない状態になります。
また、USBケーブルを抜き差しすると正常に動作するようになります。
2台のPCで同じ事象になることを確認しており、PCとの接続にUSBハブは使用しておりません。
USBケーブルの右差し・左差しやPro Microの左右入れ替えもしてみましたが変化はありませんでした。
こちらの事象について対処方法はありますでしょうか。
![20230201_234025](https://user-images.githubusercontent.com/124221527/216214576-22e1c633-06b5-4f7d-849b-f1aa4c71692f.JPG)

②USBケーブル接続していない側のLED輝度が安定しない
LED輝度MAX～２段階落としの範囲でUSBケーブルを接続していない側のLED輝度が不安定になります。
(裏表すべてのLEDが明るくなったり暗くなったりを高速で繰り返している状態)
USBケーブルの右差し・左差しどちらでも発生しています。
輝度MAXでも赤・緑・青の単色では発生しなかったり、輝度を落とすと発生しなかったりと
電圧の不足により発生しているようにも思えますが、動作としては正常でしょうか。
使用には影響がない事象ですが、上記と併せて確認させていただきたく…

**Comment by Yowkees** [comment]
いつもお世話になります。
このたびはKeyball44のご購入、誠にありがとうございます。

USBを差した状態でのPC起動時にキーボードが認識されない現象は私も遭遇します。

こちらですが、PCからUSBポートに給電されるタイミングとPCからデバイス認識するタイミングによるものと思われ、
Keyballシリーズに限らずQMK_Firmwareをベースに設計されているすべての自作キーボードで起こる問題と認識しています。

そのためQMK側もしくはマイコンのBootloader側を修正しないと直らないかと思います。

毎回抜き差しするのは大変なので、マグネット接続タイプのUSBケーブルを使用するなど、
再接続しやすく工夫して使用される方も多いです。


次にLEDのちらつきですが、
TRSコネクタの半田付けを見直していただくか、TRSケーブルを別のものに変えると改善する場合があります。

私のKeyball44でも白めの色で光量を上げると多少ちらつきますね。

これらはProMicroの電流量の限界によるものと思われます。
ProMicroの1ピンあたりの最大電流が1Aだったかと記憶していますが、
その上限近くに達しておりちらつくようです。

このような回答で恐縮ですが、ご理解いただければ幸いです。

**Comment by heppokoX** [comment]
ご回答ありがとうございました。
Lily58 ProやCorne Cherryでは起動時も正常に認識されていたため不安だったのですが、
Yowkees様の環境でも発生するとのことで一先ず安心できました。
調べてみると同事象についてファームウェアの編集にて対応されている方もいるようですので、
こちらでも改善できないか試してみたいと思います。

また、LED輝度についてもご確認ありがとうございました。
TRSコネクタのはんだ確認やケーブル交換を試みましたが変化がなかったため
ご回答通りの状態なのかと思います。
省エネ気味の運用でいきたいと思います。

**Comment by Yowkees** [comment]
そうなのですか！情報ありがとうございます。
私が３年ほど前に使用していたCorne chocolateも同症状でしたので、仕方のない現象と諦めていましたが、
ファームウェアの編集で改善できる可能性があるのですね。

もし改善できましたら、PRいただければ大変助かります。

LEDについてもご検討ありがとうございます。調整して使っていただければ幸いです。

今後とも何卒よろしくお願いいたします。

**Comment by heppokoX** [comment]
お世話になっております。
初心者なりに弄った結果、とりあえず個人的に満足のいく状態になりましたので報告させていただきます。

PC起動時の認識については`SPLIT_USB_DETECT`の定義無効化で改善した例があり、
実際にconfig.h内の下記箇所をコメントアウトとしたところPC起動時に差し替えなく認識するようになりました。

```
//#define SPLIT_USB_DETECT
//#define SPLIT_USB_TIMEOUT      500
```

左右差し替えても認識し、それぞれのキーマップで入力できるようでしたので動作に問題はないようです。
ただ、一部Pro Microの電源関連の対策かと思われますので使えるかは個々の環境に依る気がしますね…

また、LED輝度についてもconfig.h内の`RGBLIGHT_LIMIT_VAL`の値を150から125に下げることで
MAX輝度でもちらつきを無くすことができました。

`#        define RGBLIGHT_LIMIT_VAL  125 // limitated for power consumption`

私の環境でしか状態確認できていませんが、初期値はもう少し下げてもいいのではという所感です。
